### PR TITLE
Generate code for [validate]

### LINKED ISSUE #38: Generate validation code for validate attributes
FacilityApi/Facility#27 adds `[validate(...)]` as a mechanism for specifying validation criteria for use in code generation. Generated code should include this validation criteria.

### PR DESCRIPTION
**Pull request by jzbrooks**
Resolves #38

### CODE REVIEW

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs [hunk 1]
```diff
@@ -356,11 +375,80 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
...
+											case ServiceTypeKind.Int64:
+											{
+												GenerateRangeCheck(code, propertyName, validation.ValueRange!, "L");
+												break;
+											}
+											case ServiceTypeKind.Decimal:
+											{
+												GenerateRangeCheck(code, propertyName, validation.ValueRange!, "M");
+												break;
+											}
```
Comment (jzbrooks) [inline src/Facility.CodeGen.CSharp/CSharpGenerator.cs:434]: I understand there to be some implicit conversions in comparisons of numeric values (like int and long). Always adding the suffix seems more consistent to me, but let me know if you have a better way to do this.

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs [hunk 2]
```diff
@@ -356,11 +375,80 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
...
+											case ServiceTypeKind.String:
+											{
+												var validRange = validation.LengthRange;
+												if (validRange != null)
+													GenerateRangeCheck(code, $"{propertyName}.Length", validRange);
+
+												var validPattern = validation.RegexPattern;
+												if (validPattern != null)
+												{
+													code.WriteLine($"if (!System.Text.RegularExpressions.Regex.IsMatch({propertyName}, \"{validPattern}\")");
```
Comment (ejball) [inline src/Facility.CodeGen.CSharp/CSharpGenerator.cs:407]: The pattern needs to be escaped. Use `CSharpUtility.CreateString`.

Also, we shouldn't rely on the [global regex cache](https://docs.microsoft.com/en-us/dotnet/standard/base-types/compilation-and-reuse-in-regular-expressions#the-regular-expressions-cache), which means we should create a `static readonly Regex` for each pattern.

Also, feel free to add `System.Text.RegularExpressions` to the usings.

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs (hunk 1 above)
Comment (ejball) [reply src/Facility.CodeGen.CSharp/CSharpGenerator.cs:434]: Seems reasonable.

File: tests/Facility.CodeGen.CSharp.UnitTests/ValidationTests.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+			var definition = "[csharp] service TestApi { method do { [validate(length: 10..)] password: string; }: {} }";
+			var requestDtoFile = GetGeneratedFile(definition, "DoRequestDto.g.cs");
+
+			StringAssert.Contains("if (Password.Length < 10)", requestDtoFile.Text);
+		}
+
+		[Test]
+		public void GeneratesStringPattern()
+		{
+			var definition = @"[csharp] service TestApi { method do { [validate(regex: ""[0-9]{4}"")] pin: string; }: {} }";
```
Comment (ejball) [inline tests/Facility.CodeGen.CSharp.UnitTests/ValidationTests.cs:61]:
```suggestion
			var definition = @"[csharp] service TestApi { method do { [validate(regex: ""^[0-9]{4}$"")] pin: string; }: {} }";
```
I'd add anchors so that the test doesn't cause confusion on how validation works.

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs (hunk 2 above)
Comment (jzbrooks) [reply src/Facility.CodeGen.CSharp/CSharpGenerator.cs]: I'm curious. Does the cache involve some serious overhead in practice? Or is it just that LRU evictions could cause sneaky (and seriously non-deterministic) performance issues?

FWIW, My original thought was to avoid unused using warnings in generated code, assuming that kind of warning was something generated code should be concerned about. What are your thoughts on generated code style? Is the bar just generating reasonable looking code, or are there specifics beyond that?

Comment (ejball) [reply src/Facility.CodeGen.CSharp/CSharpGenerator.cs]: > LRU evictions could cause sneaky (and seriously non-deterministic) performance issues

Yep.

The generated code already has unused usings, so you wouldn't be adding a new issue. Generated code should be exempt from code style warnings.

I like generating code that's as normal-looking as possible.

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs [hunk 4]
```diff
@@ -298,6 +316,18 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
...
 
+						var regexFields = fieldInfos.Where(x => x.Validation?.RegexPattern != null).ToList();
+						if (regexFields.Count != 0)
+						{
+							code.WriteLine();
+							foreach (var fieldInfo in regexFields)
+							{
+								var propertyName = context.GetFieldPropertyName(fieldInfo);
+								var validPattern = fieldInfo.Validation!.RegexPattern!;
+								code.WriteLine($"private static readonly Regex s_Valid{propertyName}Pattern = new Regex({CSharpUtility.CreateString(validPattern)});");
```
Comment (ejball) [inline src/Facility.CodeGen.CSharp/CSharpGenerator.cs:327]:
```suggestion
								code.WriteLine($"private static readonly Regex s_valid{propertyName}Regex = new Regex({CSharpUtility.CreateString(validPattern)}, RegexOptions.CultureInvariant);");
```

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs [hunk 5]
```diff
@@ -356,11 +388,108 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
...
+											case ServiceTypeKind.String:
+											{
+												var validRange = validation.LengthRange;
+												if (validRange != null)
+												{
+													if (validRange.Minimum != null)
+													{
+														code.WriteLine($"if ({propertyName} != null && {propertyName}.Length < {validRange.Minimum})");
+														using (code.Indent())
+															code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", \"Must be at least {validRange.Minimum}\");");
```
Comment (ejball) [inline src/Facility.CodeGen.CSharp/CSharpGenerator.cs:420]:
```suggestion
															code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", \"Length must be at least {validRange.Minimum}\");");
```

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs [hunk 6]
```diff
@@ -356,11 +388,108 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
...
+													}
+
+													if (validRange.Maximum != null)
+													{
+														code.WriteLine($"if ({propertyName} != null && {propertyName}.Length > {validRange.Maximum})");
+														using (code.Indent())
+															code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", \"Must be at most {validRange.Maximum}\");");
```
Comment (ejball) [inline src/Facility.CodeGen.CSharp/CSharpGenerator.cs:427]:
```suggestion
															code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", \"Length must be at most {validRange.Maximum}\");");
```

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs [hunk 7]
```diff
@@ -356,11 +388,108 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
...
+													}
+												}
+
+												var validPattern = validation.RegexPattern;
+												if (validPattern != null)
+												{
+													var regexField = $"s_Valid{propertyName}Pattern";
+													code.WriteLine($"if ({propertyName} != null && !{regexField}.IsMatch({propertyName}))");
+													using (code.Indent())
+														code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{propertyName}\", $\"Must match {{{regexField}}}.\");");
```
Comment (ejball) [inline src/Facility.CodeGen.CSharp/CSharpGenerator.cs:437]:
```suggestion
														code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{propertyName}\", $\"Must match regular expression: {regexField}\");");

```

Brackets feel like an unusual delimiter for a regular expression.

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs [hunk 8]
```diff
@@ -356,11 +388,108 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
...
+											case ServiceTypeKind.Bytes:
+											case ServiceTypeKind.Array:
+											case ServiceTypeKind.Map:
+											{
+												var range = validation.CountRange!;
+												if (range.Minimum != null)
+												{
+													code.WriteLine($"if ({propertyName} != null && {propertyName}.Count < {range.Minimum})");
+													using (code.Indent())
+														code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", \"Must be at least {range.Minimum}\");");
```
Comment (ejball) [inline src/Facility.CodeGen.CSharp/CSharpGenerator.cs:473]:
```suggestion
														code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", \"Count must be at least {range.Minimum}\");");
```

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs [hunk 9]
```diff
@@ -356,11 +388,108 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
...
+												}
+
+												if (range.Maximum != null)
+												{
+													code.WriteLine($"if ({propertyName} != null && {propertyName}.Count > {range.Maximum})");
+													using (code.Indent())
+														code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", \"Must be at most {range.Maximum}\");");
```
Comment (ejball) [inline src/Facility.CodeGen.CSharp/CSharpGenerator.cs:480]:
```suggestion
														code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", \"Count must be at most {range.Maximum}\");");
```

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs [hunk 10]
```diff
@@ -269,6 +286,7 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
 				{
 					"System",
 					"System.Collections.Generic",
+					"System.Text.RegularExpressions",
```
Comment (ejball) [inline src/Facility.CodeGen.CSharp/CSharpGenerator.cs:289]: Is it unfortunate that the new generator causes every file to be modified... I should have thought of that.

One simple fix would be to move `regexFields` up here and only add it if there are any regex patterns. `WriteUsings` sorts the namespaces.

File: conformance/ConformanceApi.fsd [hunk 11]
```diff
@@ -121,10 +122,11 @@ service ConformanceApi
 	data Widget
 	{
 		/// A unique identifier for the widget.
+		[validate(value: 0..)]
 		id: int32;
 
 		/// The name of the widget.
-		[required]
+		[required, validate(regex: "[_a-zA-Z0-9]+")]
```
Comment (ejball) [inline conformance/ConformanceApi.fsd:129]:
```suggestion
		[required, validate(regex: "^[_a-zA-Z0-9]+$")]
```

File: src/Facility.ConformanceApi/GetWidgetRequestDto.g.cs [hunk 12]
```diff
@@ -57,6 +57,9 @@ public override bool Validate(out string? errorMessage)
 			if (Id == null)
 				return ServiceDataUtility.GetRequiredFieldErrorMessage("id");
 
+			if (Id != null && Id < 0)
+				return ServiceDataUtility.GetInvalidFieldErrorMessage("Id", "Must be at least 0");
```
Comment (ejball) [inline src/Facility.ConformanceApi/GetWidgetRequestDto.g.cs:61]:
```suggestion
				return ServiceDataUtility.GetInvalidFieldErrorMessage("id", "Must be at least 0.");
```

As on line 58 above, field names in error messages should use FSD casing.

Also, like other error messages, these should use sentence punctuation.

**Review by ejball — CHANGES_REQUESTED** [review]
Hopefully we can add an example of count and of length validation to the conformance API. Also it would be good to test `..100` and `1..100` (or some such).

**Comment by jzbrooks** [comment]
Sure thing. Wasn't sure if every use case was represented there. They'll be quick to add.

File: conformance/ConformanceApi.fsd (hunk 11 above)
Comment (jzbrooks) [reply conformance/ConformanceApi.fsd:129]: Oh right. 👍

Comment (jzbrooks) [reply conformance/ConformanceApi.fsd:129]: Should we just add the anchors implicitly? Is there ever a case where you might want to only match part of a string to determine validity?

Comment (ejball) [reply conformance/ConformanceApi.fsd]: I'm fine sticking with OpenAPI behavior here, which doesn't have implicit anchors. I suppose one might want to enforce a prefix or a suffix.

File: conformance/ConformanceApi.fsd [hunk 13]
```diff
@@ -212,8 +212,8 @@ service ConformanceApi
 	{
 		[http(from: query), required] query: string;
 		[required] normal: string;
+		[validate(count: 5)] fiveWidgets: Widget[];
 		widget: Widget;
-		widgets: Widget[];
```
Comment (ejball) [inline conformance/ConformanceApi.fsd:210]: We should avoid breaking changes to the conformance API. See [ConformanceTests.json](https://github.com/FacilityApi/FacilityCSharp/blob/master/conformance/ConformanceTests.json) for how it is used.

How about a new optional field: `[validate(count: 2)] point: double[];`

File: conformance/ConformanceApi.fsd [hunk 14]
```diff
@@ -108,7 +108,7 @@ service ConformanceApi
 	method getWidgetBatch
 	{
 		/// The IDs of the widgets to return.
-		[http(from: body), required]
+		[http(from: body), required, validate(count: ..100)]
```
Comment (ejball) [inline conformance/ConformanceApi.fsd:111]: A violation will be easier to test if the maximum count is smaller, e.g. `..10`. (My bad.)

**Review by ejball — CHANGES_REQUESTED** [review]

File: conformance/ConformanceApi.fsd (hunk 14 above)
Comment (jzbrooks) [reply conformance/ConformanceApi.fsd:111]: On my machine I have a test that uses `Enumerable.Range(0, 101)`, but I can change it.
The big number and weird initialization doesn't make the test anymore true.

File: tests/Facility.ConformanceApi.UnitTests/DtoValidationTests.cs [hunk 15]
```diff
@@ -182,7 +182,7 @@ public void RequiredWidgetNameMissing()
 		public void RequiredWidgetsNameMissing()
 		{
 			var dto = CreateRequiredRequest();
-			dto.Widgets = new[] { CreateWidget(), new WidgetDto() };
+			dto.Widgets = new[] { CreateWidget(), new(), new(), new(), new() };
```
Comment (ejball) [inline tests/Facility.ConformanceApi.UnitTests/DtoValidationTests.cs:185]: Unnecessary change.

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs [hunk 16]
```diff
@@ -356,11 +391,108 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
...
+													}
+												}
+
+												var validPattern = validation.RegexPattern;
+												if (validPattern != null)
+												{
+													var regexField = $"s_valid{propertyName}Regex";
+													code.WriteLine($"if ({propertyName} != null && !{regexField}.IsMatch({propertyName}))");
+													using (code.Indent())
+														code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", $\"Must match regular expression: {{{regexField}}}.\");");
```
Comment (ejball) [inline src/Facility.CodeGen.CSharp/CSharpGenerator.cs:440]:
```suggestion
														code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", $\"Must match regular expression: {{{regexField}}}\");");
```

Nit: I omit the trailing period when I use the colon pattern, which I use when I want to display a string without worrying about delimiters.

**Review by ejball — APPROVED** [review]

**Comment by jzbrooks** [comment]
@ejball, how much value do you think 2ab04e7996666b8f2a02f668da22cf5e7332d9e5 adds to the test suite? Validation has pretty good coverage in unit tests. I can revert the commit if it isn't valuable or add a few more if it is.

**Comment by ejball** [comment]
@jzbrooks, looks good to me! Maybe `createWidgetNameTooLong` and `checkQueryInvalidAnswer` to round things out?

**Comment by ejball** [comment]
Let me know if you need any further review from me. Otherwise feel free to merge when you are ready.

**Comment by jzbrooks** [comment]
The only other question is shipping a package in this PR. Without objection, I'll rev the minor version and update the change log.
